### ISSUE TITLE
Adding Multi-Similarity Loss to TFA losses

### ISSUE THREAD

**Issue by shun-lin** [tensorflow/addons#1017]
**Describe the feature and the current behavior/state.**

The paper's abstract:

<em>
A family of loss functions built on pair-based computation have been proposed in the literature which provide a myriad of solutions for deep metric learning. In this paper, we provide a general weighting framework for understanding recent pair-based loss functions. Our contributions are three-fold: (1) we establish a General Pair Weighting (GPW) framework, which casts the sampling problem of deep metric learning into a unified view of pair weighting
through gradient analysis, providing a powerful tool for understanding recent pair-based loss functions; (2) we show that with GPW, various existing pair-based methods can be
compared and discussed comprehensively, with clear differences and key limitations identified; (3) we propose a new loss called multi-similarity loss (MS loss) under the GPW,
which is implemented in two iterative steps (i.e., mining and weighting). This allows it to fully consider three similarities for pair weighting, providing a more principled approach for collecting and weighting informative pairs. Finally, the proposed MS loss obtains new state-of-the-art performance on four image retrieval benchmarks, where it outperforms the most recent approaches, such as ABE and HTL, by a large margin, e.g., 60.6% → 65.7% on CUB200, and 80.9% → 88.0% on In-Shop Clothes Retrieval dataset at Recall@1.  </em>

**Relevant information**
- Are you willing to contribute it (yes/no): yes
- Are you willing to maintain it going forward? (yes/no): yes
- Is there a relevant academic paper? (if so, where): yes, http://openaccess.thecvf.com/content_CVPR_2019/papers/Wang_Multi-Similarity_Loss_With_General_Pair_Weighting_for_Deep_Metric_Learning_CVPR_2019_paper.pdf

- Is there already an implementation in another framework? (if so, where): yes 

In Pytorch (original):
https://github.com/MalongTech/research-ms-loss

In first reimplementation in TensorFlow (1.x):
https://github.com/geonm/tf_ms_loss

Small/quick fixes on the above code and wrap it in `keras.loss`
https://github.com/shun-lin/multi-similarity-loss-tensorflow/blob/master/Multi_Similarity_Loss_Tensorflow.ipynb

- Was it part of tf.contrib? (if so, where): No

**Which API type would this fall under (layer, metric, optimizer, etc.)**
Loss

**Who will benefit with this feature?**
People who want to try out multi-similarity loss for their metric learning task.

**Comment by seanpmorgan** [comment]
Looks like a nice addition to me.

**Comment by ashutosh1919** [comment]
@shun-lin and @seanpmorgan, please review PR #1679

**Comment by sarim-zafar** [comment]
Any update on this?

**Comment by seanpmorgan** [comment]
TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
